azure-iot-device/azure/iot/device/iothub/pipeline/mqtt_topic_iothub.py

```diff
@@ -46,22 +50,65 @@ def get_input_topic_for_subscribe(device_id, module_id):
     return _get_topic_base(device_id, module_id) + "/inputs/#"
 
 
+def get_method_topic_for_subscribe():
+    """
+    :return: The topic for ALL incoming methods. It is of the format
+    "$iothub/methods/POST/#"
+    """
+    return "$iothub/methods/POST/#"
+
+
+def get_twin_response_topic_for_subscribe():
+    """
+    :return: The topic for ALL incoming twin responses. It is of the format
+    "$iothub/twin/res/#"
+    """
+    return "$iothub/twin/res/#"
+
+
+def get_twin_patch_topic_for_subscribe():
+    """
+    :return: The topic for ALL incoming twin patches. It is of the format
+    "$iothub/twin/PATCH/properties/desired/#
+    """
+    return "$iothub/twin/PATCH/properties/desired/#"
+
+
+def get_telemetry_topic_for_publish(device_id, module_id):
+    """
+    return the topic string used to publish telemetry
+    """
+    return _get_topic_base(device_id, module_id) + "/messages/events/"
+
+
 def get_method_topic_for_publish(request_id, status):
     """
     :return: The topic for publishing method responses. It is of the format
     "$iothub/methods/res/<status>/?$rid=<requestId>
+
+    Note that all inputs MUST be strings (not ints!)
     """
     return "$iothub/methods/res/{status}/?$rid={request_id}".format(
-        status=urllib.parse.quote_plus(status), request_id=urllib.parse.quote_plus(request_id)
+        status=urllib.parse.quote(status, safe=""),
+        request_id=urllib.parse.quote(request_id, safe=""),
     )
 
 
-def get_method_topic_for_subscribe():
+# NOTE: Consider splitting this into separate logic for Twin Requests / Twin Patches
+# This is the only method that is shared. Would probably simplify code if it was split.
+# Please consider refactoring.
+def get_twin_topic_for_publish(method, resource_location, request_id):
     """
-    :return: The topic for ALL incoming methods. It is of the format
-    "$iothub/methods/POST/#"
+    :return: The topic for publishing twin requests / patches. It is of the format
+    "$iothub/twin/<method><resourceLocation>?$rid=<requestId>
+
+    Note that all inputs MUST be strings (not ints!)
```

Comment:
>not ints [](start = 42, length = 8)

are these being specified for the request ids ? should not these be then for device ids and module ids also ?

Reply:
The specification here is due to the status code. Something like 200, 501, 404, etc. Naturally those are probably numbers instead of strings.

Better idea is probably just to type convert in here though, I think I'll change that.

---
In reply to: [399725688](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399725688) [](ancestors = 399725688)